File: playwright-csharp/tests/Support/Hooks.cs
using Reqnroll;
using Reqnroll.Infrastructure;
using TestTracks.Playwright.CSharp.Specs.Reporting;

namespace TestTracks.Playwright.CSharp.Specs.Support;

[Binding]
public sealed class Hooks
{
    private readonly ScenarioState _state;
    private readonly ScenarioContext _scenarioContext;
    private readonly ReqnrollOutputHelper _outputHelper;

    public Hooks(
        ScenarioState state,
        ScenarioContext scenarioContext,
        ReqnrollOutputHelper outputHelper)
    {
        _state = state;
        _scenarioContext = scenarioContext;
        _outputHelper = outputHelper;
    }

    [AfterScenario]
    public async Task AfterScenarioAsync()
    {
        var failed = _scenarioContext.TestError is not null;
        var evidence = new EvidenceSupport(_state, _outputHelper);
        var teardownFailures = new List<Exception>();

        try
        {
            await evidence.AttachAsync(failed);
        }
        catch (Exception ex)
        {
            teardownFailures.Add(new InvalidOperationException("Evidence capture failed.", ex));
        }

        try
        {
            teardownFailures.AddRange(await _state.Cleanup.RunAsync());
        }
        catch (Exception ex)
        {
            teardownFailures.Add(new InvalidOperationException("Cleanup failed before all registered cleanup actions could be completed.", ex));
        }

        try
        {
            await _state.Resources.DisposeAsync();
        }
        catch (Exception ex)
        {
            teardownFailures.Add(new InvalidOperationException("Resource disposal failed.", ex));
        }

        if (teardownFailures.Count == 0)
        {
            return;
        }

        if (failed)
        {
            WriteTeardownFailures(teardownFailures);
            return;
        }

        throw new AggregateException(teardownFailures);
    }

    private void WriteTeardownFailures(IReadOnlyList<Exception> failures)
    {
        _outputHelper.WriteLine("== Teardown failures ==");

        foreach (var failure in failures)
        {
            _outputHelper.WriteLine(failure.ToString());
        }
    }
}

File: playwright-java/src/test/java/testtracks/support/Hooks.java
package testtracks.support;

import io.cucumber.java.After;
import io.cucumber.java.Scenario;
import java.util.ArrayList;
import testtracks.reporting.EvidenceSupport;

public final class Hooks {
  private final ScenarioState state;

  public Hooks(ScenarioState state) {
    this.state = state;
  }

  @After
  public void afterScenario(Scenario scenario) {
    var failed = scenario.isFailed();
    var teardownFailures = new ArrayList<Exception>();

    try {
      new EvidenceSupport(state, scenario).attach(failed);
    } catch (Exception ex) {
      teardownFailures.add(new IllegalStateException("Evidence capture failed.", ex));
    }

    try {
      teardownFailures.addAll(state.cleanup().run());
    } catch (Exception ex) {
      teardownFailures.add(
          new IllegalStateException(
              "Cleanup failed before all registered cleanup actions could be completed.", ex));
    }

    try {
      state.resources().close();
    } catch (Exception ex) {
      teardownFailures.add(new IllegalStateException("Resource disposal failed.", ex));
    }

    if (teardownFailures.isEmpty()) {
      return;
    }

    if (failed) {
      scenario.log("== Teardown failures ==");
      teardownFailures.forEach(failure -> scenario.log(failure.toString()));
      return;
    }

    throw new IllegalStateException("Teardown failed.", teardownFailures.getFirst());
  }
}
